pennylane_lightning/core/src/measurements/MeasurementsBase.hpp

```diff
@@ -111,6 +135,167 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
+     * @param shot_range The range of samples to use. If it's empty, all samples
+     * are used.
+     *
+     * @return Expectation value with respect to the given observable.
+     */
+    auto expval(const Observable<StateVectorT> &obs, const size_t &num_shots,
+                const std::vector<size_t> &shot_range) -> PrecisionT {
+        PrecisionT result = 0;
+        std::vector<size_t> short_range = {};
+
+        if (obs.getObsName().find("SparseHamiltonian") != std::string::npos) {
+            PL_ABORT("For SparseHamiltonian Observables, expval calculation is "
+                     "not supported by shots");
+        } else if (obs.getObsName().find("Hermintian") != std::string::npos) {
+            PL_ABORT("For Hermintian Observables, expval calculation is not "
+                     "supported by shots");
+        } else if (obs.getObsName().find("Hamiltonian") != std::string::npos) {
+            auto coeffs = obs.getCoeffs();
+            for (size_t obs_term_idx = 0; obs_term_idx < coeffs.size();
+                 obs_term_idx++) {
+                auto obs_samples =
+                    samples(obs, num_shots, shot_range, obs_term_idx);
+                size_t num_elements = 0;
+                PrecisionT result_per_term = 0.0;
+                for (int element : obs_samples) {
+                    result_per_term += element;
+                    num_elements++;
+                }
+                result += coeffs[obs_term_idx] * result_per_term / num_elements;
+            }
+        } else {
+            auto obs_samples = samples(obs, num_shots, shot_range);
+            size_t num_elements = 0;
+            for (int element : obs_samples) {
+                result += element;
+                num_elements++;
+            }
+            result = result / num_elements;
+        }
+
+        return result;
+    }
+
+    /**
+     * @brief Return samples of a observable
+     *
+     * @param obs The observable to sample
+     * @param num_shots Number of shots used to generate samples
+     * @param shot_range The range of samples to use. If it's empty, all samples
+     * are used.
+     * @param bin_size  Divides the shot range into bins of size ``bin_size``,
+     * and returns the measurement statistic separately over each bin.
+     * @param counts Whether count("True") or raw samples ("False") should be
+     * retruned
+     *
+     * @return std::vector<size_t> samples in std::vector
+     */
+    auto samples(const Observable<StateVectorT> &obs, const size_t &num_shots,
+                 const std::vector<size_t> &shot_range,
+                 const size_t term_idx = 0,
+                 [[maybe_unused]] size_t bin_size = 0,
+                 [[maybe_unused]] bool counts = false) {
+        std::vector<size_t> obs_wires; // for Hamiltonian this obs_wires should
+                                       // be calculated based on terms
+        const size_t num_qubits = _statevector.getTotalNumQubits();
```

Comment:
Why not using `getNumQubits`?
```suggestion
        const size_t num_qubits = _statevector.getNumQubits();
```

Reply:
See above!